src/tdastro/base_models.py

```diff
@@ -32,20 +53,85 @@ def __init__(self, host=None, ra=None, dec=None, distance=None, **kwargs):
            Any additional keyword arguments.
         """
         self.host = host
+        self.effects = []
+        self.setters = {}
+
+        # Set RA, dec, and distance from the parameters.
+        self.add_parameter("ra", ra)
+        self.add_parameter("dec", dec)
+        self.add_parameter("distance", distance)
+
+    def add_parameter(self, name, value=None, required=False, **kwargs):
+        """Add a single parameter to the PhysicalModel. Checks multiple sources
+        in the following order:
+            1. Manually specified ``value``
+            2. An entry in ``kwargs``
+            3. An entry in the object's host.
+            4. ``None``
+        Sets an initial value for the attribute based on the given information.
+
+        Parameters
+        ----------
+        name : `str`
+            The parameter name to add.
+        value : any, optional
+            The information to use to set the parameter. Can be a constant
+            or a function.
+        required : `bool`
+            Fail if the parameter is set to ``None``.
+        **kwargs : `dict`, optional
+           All other keyword arguments, possibly including the parameter setters.
 
-        # Set RA, dec, and distance from the given value or, if it is None and
-        # we have a host, from the host's value.
-        self.ra = ra
-        self.dec = dec
-        self.distance = distance
-        if ra is None and host is not None:
-            self.ra = host.ra
-        if dec is None and host is not None:
-            self.dec = host.dec
-        if distance is None and host is not None:
-            self.distance = host.distance
+        Raises
+        ------
+        Raise a ``KeyError`` if there is a parameter collision.
+        Raise a ``ValueError`` if the parameter is required, but set to None.
+        """
+        if hasattr(self, name):
```

Comment:
```suggestion
        if hasattr(self, name) and getattr(self, name) is not None:
```
I wonder about adding this in for cases where we've set the attribute as None, but would like to override this (especially for attrs like `ra`, `dec`, and `distance`, that default to None in the base model).

As is, the following will raise a KeyError:
```
model = StepSource(
        brightness=_sample_brightness,
        t_start=0.0,
        t_end=_sample_end,
        magnitude=100.0,
        duration=5.0,
    )
    model.add_parameter('ra', 0.0)
```

However, I do not know how much we would want to encourage post-hoc setting of fundamental attributes like this.